### ISSUE TITLE
一些对于项目代码的疑惑

### ISSUE THREAD

**Issue by dornd2000** [SuperMHP/GUPNet_Plus#10]
首先，很感谢您的工作。
然后，由于我在新版的mmdet3d上对您的代码进行了迁移，关于代码中存在一些问题没弄明白，希望您能帮助解答：

1. 关于您添加的coord_map部分，为什么是角度上的平均？
coord_maps = torch.cat([torch.cat([coords_in_camera_coord[:,0:1]+i*(coords_in_camera_coord[:,2:3]-coords_in_camera_coord[:,0:1])/6 for i in range(7)],-1).unsqueeze(1).repeat([1,7,1]).unsqueeze(1),
                                        torch.cat([coords_in_camera_coord[:,1:2]+i*(coords_in_camera_coord[:,3:4]-coords_in_camera_coord[:,1:2])/6 for i in range(7)],-1).unsqueeze(2).repeat([1,1,7]).unsqueeze(1)],1)
这是您的代码，但我的理解是您在做一个类似射线角度上的绝对位置信息，但为什么是先获得左上和右下的(X/Z, Y/Z)角度，再均匀分割成7*7的矩阵；我的想法是先获得7*7共49个中心再转为角度，这样可能更对应。

2. 关于 $\beta\_NLL$ 的实现代码，什么是在平均后再添加 beta？
您重写了 UncertainL1Loss 类。这里 uncertain_l1_loss 返回是一个标量，但sigma.detach()是一个矩阵，mmdet3d的逻辑应该是：loss如果是矩阵会自动算平均。也就是，您先计算了一般的Laplacian损失，然后相当于添加了一个受sigma.detach()影响的标量？这里我觉得和您论文中写的不太一致，且这可能使得损失受到同一批数据的影响，所以我尝试将sigma.detach()放入uncertain_l1_loss 但发现模型效果有所下降，请问您为什么这么设计？以下是您的代码：
loss_bbox = torch.exp(self.beta*sigma.detach()) * (self.loss_weight)**(-self.beta) * \
            (self.loss_weight * uncertain_l1_loss(
            pred,
            target,
            weight,
            sigma=sigma,
            alpha=self.alpha,
            reduction=reduction,
            avg_factor=avg_factor))

3. 以及一个小点，您为什么选择Adam策略而不是AdamW策略，因为我在迁移后效果虽然和您的相近，但也有少量的差别，所以我仔细检查了和您代码的差别，并进行了调整。应用AdamW后，mAP在kitti valdataset easy上最高大致是28+但没到29，但遵循您的处理应用Adam效果会有下滑（其他配置基本和您的一致），这也让我更加疑惑。

期待您的回复。

**Comment by dornd2000** [comment]
感谢您的回答，我的描述不太清楚，我重新说明一下。
1. 第一个问题。从代码看您先计算了boxes 左上和右下点的在相机坐标系下的相关坐标 $(X_1/Z_1, Y_1/Z_1)$ 和 $(X_2/Z_2, Y_2/Z_2)$（因为这个坐标是射线相关的坐标所以我之前称其为角度），然后您将其均匀的划分成7*7的尺寸以配合您的roi输出，您在X/Z和Y/Z坐标上的划分是 $[0, 1\times\frac76, 1\times\frac76,2\times\frac76,3\times\frac76,4\times\frac76,5\times\frac76,6\times\frac76]$，当然这还需要乘上 左上点和右下点的差 并加上 左上的位移量，但我的理解中您的这些坐标和图像上 roi 的均匀划分其实并没有对齐。
在我的理解中，如果寻找和roi对齐的划分，应该先获取每个roi划分小框的中心坐标，即 $[0.5,1.5,2.5,3.5,4.5,5.6,6.5]$（这里也需要乘以差和放缩，但不同的是这是先在图像上计算的），然后利用calib转移回相机坐标系下的 $(X/Z,Y/Z)$。

您的设计和我的想法的代码如下：
`
#这是您的代码，我稍微调整过一些但逻辑一致
if roi_calibs.shape[-1]==4: 
                    coords_in_camera_coord = torch.cat([(roi_calibs.inverse() @ torch.cat([bbox2d_maps_ori[:,:2],bbox2d_maps_ori.new_ones([num_masked_bin,2])],-1).unsqueeze(-1))[:,:2, 0],
                                                    (roi_calibs.inverse() @ torch.cat([bbox2d_maps_ori[:,2:],bbox2d_maps_ori.new_ones([num_masked_bin,2])],-1).unsqueeze(-1))[:,:2, 0]],-1)     # [N*2, 2]
                else:
                    coords_in_camera_coord = torch.cat([(roi_calibs.inverse() @ torch.cat([bbox2d_maps_ori[:,:2],bbox2d_maps_ori.new_ones([num_masked_bin,1])],-1).unsqueeze(-1))[:,:2, 0],
                                                    (roi_calibs.inverse() @ torch.cat([bbox2d_maps_ori[:,2:],bbox2d_maps_ori.new_ones([num_masked_bin,1])],-1).unsqueeze(-1))[:,:2, 0]],-1)
                coord_maps = torch.cat([torch.cat([coords_in_camera_coord[:,0:1]+i*(coords_in_camera_coord[:,2:3]-coords_in_camera_coord[:,0:1])/(self.roi_output_size-1) for i in range(self.roi_output_size)],-1).unsqueeze(1).repeat([1,self.roi_output_size,1]).unsqueeze(1),
                                        torch.cat([coords_in_camera_coord[:,1:2]+i*(coords_in_camera_coord[:,3:4]-coords_in_camera_coord[:,1:2])/(self.roi_output_size-1) for i in range(self.roi_output_size)],-1).unsqueeze(2).repeat([1,1,self.roi_output_size]).unsqueeze(1)],1)
                                        
#这是我的想法
def get_mesh_grid(bbox2d, size):
    '''
    bbox2d: (N, 4) xyxy
    size: int
    return: (N, 2, size, size) xy
    '''
    bbox2d = bbox2d.clone().detach()
    norm_coord = 0.5+torch.arange(0, size, dtype=torch.float32, device=bbox2d.device)
    grid_y, grid_x = torch.meshgrid(norm_coord, norm_coord, indexing='ij')
    
    x1 = bbox2d[:, 0:1].unsqueeze(-1) # (N, 1, 1)
    y1 = bbox2d[:, 1:2].unsqueeze(-1) # (N, 1, 1)
    roi_w = bbox2d[:, 2:3].unsqueeze(-1) - x1 # (N, 1, 1)
    roi_h = bbox2d[:, 3:4].unsqueeze(-1) - y1 # (N, 1, 1)

    # 计算步长 (每个网格代表的像素距离)
    step_w = roi_w / size
    step_h = roi_h / size

    # grid_x 是 (H, W)，step_w 是 (N, 1, 1) -> 广播为 (N, H, W)
    pixel_x = x1 + grid_x.unsqueeze(0) * step_w
    pixel_y = y1 + grid_y.unsqueeze(0) * step_h

    # coord_maps 形状 -> (N, 2, H, W)
    coord_maps = torch.stack([pixel_y, pixel_x], dim=1)
    return coord_maps

coord_maps = get_mesh_grid(bbox2d_maps_ori, self.roi_output_size)
N, _, H, W = coord_maps.shape
if roi_calibs.shape[-1]==4:
    coord_maps_h = torch.cat([coord_maps, torch.ones_like(coord_maps[:, :2, :, :], device=DEVICE)], dim=1)
else:
    coord_maps_h = torch.cat([coord_maps, torch.ones_like(coord_maps[:, :1, :, :], device=DEVICE)], dim=1)
coord_maps = (roi_calibs.inverse() @ coord_maps_h.view(N, -1, H*W))[:,:2,:].view(N, 2, H, W)`

2. 第二个问题。这里我先说明一些我认为的参数的维度和类及函数。UncertainL1Loss是一个类，uncertain_l1_loss是forward其中调用的一个函数，三个关键的变量，pred，target 和 sigma 的维度都是 (N, 1)，N 是一个batch中所有已标注物体的数量，这个数量并不固定。我的问题是：uncertain_l1_loss这个函数默认计算“mean”也就是会返回一个标量，但您会再乘以sigma.detach()**0.5，又将其变为 (N,1) 的矩阵，所以您的loss返回的是 (N,1) 的矩阵，虽然mmdet3d对于这种情况再后期应该会自动计算均值。
我对您论文中的理解相当于，uncertain_l1_loss的reduction为”none“，即 uncertain_l1_loss的返回是一个(N,1)的矩阵，再与sigma.detach()**0.5相乘。
”添加了一个受sigma.detach()影响的标量“我这句话的意思是，您这里的：
torch.exp(self.beta*sigma.detach()) * (self.loss_weight)**(-self.beta) *(self.loss_weight * uncertain_l1_loss
实际相当于：
torch.exp(self.beta*sigma.detach()).mean() * (self.loss_weight)**(-self.beta) *(self.loss_weight * uncertain_l1_loss
torch.exp(self.beta*sigma.detach()).mean()会导致：对一个物体的预测会受到同一批其他物体的影响，除非这是您刻意而为。

期待您的回复。

**Comment by dornd2000** [comment]
感谢您的回复。

1. 确实如您所说，coord maps 确实可能存在一个映射关系。
2. 关于 $\beta\_NLL$，我有些新的问题想要请教。您提到“Sigma 仍然起到了对困难样本（组）降低权重的正则化作用”。但就我的理解，在损失上您是乘上 $(\frac{\sigma}{\sqrt2})^{0.5}$，对于困难样本我更倾向于sigma较大，所以这不是增大了权重吗？以及就我的实验来看，sigma 在训练晚期应该是很容易 <1 的（表现为loss低于0，而只有log \sigma能提供负值），这时候 $(\frac{\sigma}{\sqrt2})^{0.5}$ 应该 <1，这应该导致简单样本在后期反而会被抑制进一步的优化。但如果我的实验没有出错，您的 $\beta\_NLL$ 确实带来的很大的优化效果，这也让我很困惑。
3. “Batch-based Sigma 仍然起到了对困难样本（组）降低权重的正则化作用”这句话可能也没错，但我的理解是，多数的简单样本拉低了困难样本原本的权重（因为简单样本的sigma一般确实更低），但“样本级（sample-wise）”的sigma应该无法做到这一点，而“组级（batch-wise）/粗粒度”可以。

期待您的回复。

**Comment by SuperMHP** [comment]
关于第二点，抱歉我写的文字有歧义，更准确的说法是降低“\sqrt{2}/sigma"这一项的“加权作用”，对于困难样本大sigma会使得uncertainty weighting作用显著（权重大幅下降，在训练时贡献显著下降），乘以sigma.detach()后会弱化这种weighting作用（增加其训练的贡献权重），这是原始的motivation

关于第三点，您的分析是有效的，但是结论是由于我上次文字表述不清晰因此有误差。应该是，多数简单样本的平均拉低了sigma，使得对于困难样本，其权重数值本身的增加有限（也就是detach降低weighting效应的能力比预期的sample wise要弱）。因为实际上在我的观察（准确是记忆中）中，对于kitti是很容易普遍sigma很小，但是对于nuScenes，大多数sigma都不容易小于1。这个loss的提出其实更多是为了解决nuScenes普遍sigma大导致的训练不充分。

**Comment by dornd2000** [comment]
十分感谢，我理解了。

**Comment by dornd2000** [comment]
我按照您的想法，修改成了sample级别的sigma，出现了有趣的结果。batch级别的sigma貌似带来的效果更优异，但相比标准的Laplacian Loss 确实有一定的提升，下面是在kitti val 上验证的效果，如果您感兴趣的话可以尝试分析一下。
### sample-wise Laplacian loss
Car AP40@0.70, 0.70, 0.70:
bbox AP40:93.3830, 84.4047, 77.4158
bev  AP40:36.1524, 26.5973, 23.6287
3d   AP40:26.5481, 19.3265, 16.8764
aos  AP40:92.53, 83.26, 75.85
Car AP40@0.70, 0.50, 0.50:
bbox AP40:93.3830, 84.4047, 77.4158
bev  AP40:69.8098, 52.9167, 48.0795
3d   AP40:64.3288, 48.4320, 43.6575
aos  AP40:92.53, 83.26, 75.85

### Laplacian loss
Car AP40@0.70, 0.70, 0.70:
bbox AP40:92.9372, 82.7940, 76.0180
bev  AP40:33.1220, 24.8689, 22.6995
3d   AP40:24.3531, 17.6598, 15.8351
aos  AP40:92.12, 81.52, 74.26
Car AP40@0.70, 0.50, 0.50:
bbox AP40:92.9372, 82.7940, 76.0180
bev  AP40:68.6010, 52.0412, 47.6125
3d   AP40:63.0181, 47.3059, 43.0882
aos  AP40:92.12, 81.52, 74.26

### sample-wise Laplacian loss 的函数
@weighted_loss
def uncertain_l1_loss_gup(pred, target, sigma, alpha=1.0, beta=1.0):
    assert target.numel() > 0
    assert pred.size() == target.size() == sigma.size(), 'The size of pred ' \
        f'{pred.size()}, target {target.size()}, and sigma {sigma.size()} ' \
        'are inconsistent.'
    loss = torch.abs(pred - target)
    loss = torch.exp(-sigma) * loss + alpha * sigma
    loss = torch.exp(beta*sigma.detach()) * loss
    return loss


@MODELS.register_module()
class UncertainL1Loss_gup(nn.Module):
    def __init__(self, alpha=1.0, beta=0.0, reduction='mean', loss_weight=1.0):
        super(UncertainL1Loss_gup, self).__init__()
        assert reduction in ['none', 'sum', 'mean']
        self.alpha = alpha
        self.beta = beta
        self.reduction = reduction
        self.loss_weight = loss_weight

    def forward(self,
                pred,
                target,
                sigma,
                weight=None,
                avg_factor=None,
                reduction_override=None):
        assert reduction_override in (None, 'none', 'mean', 'sum')
        reduction = (
            reduction_override if reduction_override else self.reduction)
        loss_bbox = uncertain_l1_loss_gup(
            pred,
            target,
            weight,
            sigma=sigma,
            alpha=self.alpha,
            beta=self.beta,
            reduction=reduction,
            avg_factor=avg_factor)
        loss_bbox = loss_bbox * (self.loss_weight ** (1-self.beta))
        return loss_bbox

**Comment by SuperMHP** [comment]
好的谢谢

**Comment by SuperMHP** [comment]
这里可能也有一个需要和您同步的细节，在我们的经验中，KITTI由于train的数据量以及eval/test数据量普遍较小，因此上面的结果有一定波动性，如果真的需要探究这两种策略的系统差异，可能多次的结果综合对照会更有参考价值。如果在nuScenes这样的大数据集上，也许可能会有与KITTI不同的结论。

再次感谢您的兴趣与结果，之后我们有时间的时候也会围绕此问题再深入研究，再次感谢

**Comment by dornd2000** [comment]
kitti数据集数据量小的问题确实存在，我之后会多测几次。并且如果我有时间会测下nuscenes数据集。再次感谢您的耐心解答。

**Comment by dornd2000** [comment]
您好，再次打扰，关于您gupnet++的附录中size2d，size3d和depth的分布图请问您是如何统计的，size3d我还能想到用类别均值处理，但size2d和depth的均值您是如何获得的？就是以下这张图

<img width="1079" height="287" alt="Image" src="https://github.com/user-attachments/assets/b8ccf60e-7893-47db-b665-9d22857c2290" />

**Comment by SuperMHP** [comment]
您好，在eval data上用其他实验脚本保存了对应object的h2d/h3d/depth的pred/pred uncertainty以及对应gt后，统计的。在这个repo里应该没有对应的内容，所以如果您需要相关功能的话可能需要自行编写。在 2026年4月16日，11:52，dornd2000 ***@***.***> 写道：﻿dornd2000 left a comment (SuperMHP/GUPNet_Plus#10)
您好，再次打扰，关于您gupnet++的附录中size2d，size3d和depth的分布图请问您是如何统计的，size3d我还能想到用类别均值处理，但size2d和depth的均值您是如何获得的？就是以下这张图
image.png (view on web)

**Comment by SuperMHP** [comment]
对于h2d，depth均值，是指每个box的h2d和depth的prediction，即为均值在 ***@***.*** 写道：﻿您好，在eval data上用其他实验脚本保存了对应object的h2d/h3d/depth的pred/pred uncertainty以及对应gt后，统计的。在这个repo里应该没有对应的内容，所以如果您需要相关功能的话可能需要自行编写。在 2026年4月16日，11:52，dornd2000 ***@***.***> 写道：﻿dornd2000 left a comment (SuperMHP/GUPNet_Plus#10)
您好，再次打扰，关于您gupnet++的附录中size2d，size3d和depth的分布图请问您是如何统计的，size3d我还能想到用类别均值处理，但size2d和depth的均值您是如何获得的？就是以下这张图
image.png (view on web)

**Comment by dornd2000** [comment]
您好，抱歉回的有些迟了。我明白您的意思了，十分感谢。
但还有一个问题，按照您的说法，这种统计是利用预测值和gt结合处理的，但这种统计应该用训练好的模型保存预测结果再统计的。那么对于拉普拉斯分布的假设是否会影响预测的分布，也就是说使用不同的损失函数，例如：拉普拉斯负对数损失和高斯负对数损失，是否会对最终统计的分布有干扰，亦或者说是：使用拉普拉斯损失才对应拉普拉斯分布，高斯损失可能在统计上更接近高斯分布，我并不知道您在统计的是否考虑过这个问题。当然您后面的mean error表格也能够说明拉普拉斯分布假设的效果更好。
再次感谢您的回复。

**Comment by SuperMHP** [comment]
是的，您说的是正确的，这种分布会有绑定关系，在本文中我们并没有仔细考虑该问题，但是在其他我们涉及不确定性的project中，我们做过分析。结论与您的判断一致，loss假设的分布与统计分布高度相关。高斯训练后error更接近高斯，laplace训练后error更靠近laplace。在 2026年4月19日，20:25，dornd2000 ***@***.***> 写道：﻿dornd2000 left a comment (SuperMHP/GUPNet_Plus#10)
您好，抱歉回的有些迟了。我明白您的意思了，十分感谢。
但还有一个问题，按照您的说法，这种统计是利用预测值和gt结合处理的，但这种统计应该用训练好的模型保存预测结果再统计的。那么对于拉普拉斯分布的假设是否会影响预测的分布，也就是说使用不同的损失函数，例如：拉普拉斯负对数损失和高斯负对数损失，是否会对最终统计的分布有干扰，亦或者说是：使用拉普拉斯损失才对应拉普拉斯分布，高斯损失可能在统计上更接近高斯分布，我并不知道您在统计的是否考虑过这个问题。当然您后面的mean error表格也能够说明拉普拉斯分布假设的效果更好。
再次感谢您的回复。
